src/websocket_tracker_connection.cpp

```diff
@@ -0,0 +1,490 @@
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+POSSIBILITY OF SUCH DAMAGE.
+
+*/
+
+#include "libtorrent/config.hpp" // for TORRENT_USE_RTC
+
+#if TORRENT_USE_RTC
+
+#include "libtorrent/aux_/escape_string.hpp"
+#include "libtorrent/aux_/session_settings.hpp"
+#include "libtorrent/utf8.hpp"
+#include "libtorrent/io.hpp"
+#include "libtorrent/ip_filter.hpp"
+#include "libtorrent/socket.hpp"
+#include "libtorrent/socket_io.hpp"
+#include "libtorrent/span.hpp"
+#include "libtorrent/tracker_manager.hpp"
+#include "libtorrent/websocket_tracker_connection.hpp"
+
+#include "libtorrent/aux_/disable_warnings_push.hpp"
+#include "boost/json.hpp"
+#include "libtorrent/aux_/disable_warnings_pop.hpp"
+
+#include <boost/system/system_error.hpp>
+
+#include <algorithm>
+#include <cctype>
+#include <cstdio> // for snprintf
+#include <exception>
+#include <functional>
+#include <list>
+#include <locale>
+#include <string>
+#include <string_view>
+#include <vector>
+
+namespace libtorrent {
+
+using namespace std::placeholders;
+namespace errc = boost::system::errc;
+namespace error = boost::asio::error;
+namespace json = boost::json;
+
+// Convert ISO-8859-1 (aka latin1) input to UTF-8
+std::string from_latin1(span<char const> s) {
+	std::u32string u32;
+	for(unsigned char c : s)
+		u32.push_back(c);
+	return utf32_utf8(u32);
+}
+
+// Convert UTF-8 input to ISO-8859-1 (aka latin1)
+// Throw an invalid_argument exception if it finds an unrepresentable character.
+std::string to_latin1(std::string_view sv) {
+	std::u32string u32 = utf8_utf32(sv);
+	std::string out;
+	for(char32_t cp : u32)
+	{
+		if(cp > 0xFF)
+			throw std::invalid_argument("code point out of latin1 range: " + std::to_string(cp));
+		out.push_back(static_cast<unsigned char>(cp));
+	}
+	return out;
+}
+
+// Overload for JSON strings
+std::string to_latin1(json::string_view sv) {
+      return to_latin1(std::string_view{sv.data(), sv.size()});
+}
+
+
+websocket_tracker_connection::websocket_tracker_connection(io_context& ios
+		, tracker_manager& man
+		, tracker_request const& req
+		, std::weak_ptr<request_callback> cb)
+	: tracker_connection(man, req, ios, cb)
+	  , m_io_context(ios)
+	  , m_ssl_context(ssl::context::tlsv12_client)
+{
+	queue_request(req, cb);
```

Comment:
so is, `req` and `cb` duplicated here? they are stored in the pending queue *and* in the base class?

Reply:
Yes, `req` is duplicated here. It's only for the first request though, the next ones will just be pushed in the queue and will replace the one in the base class on send.

Reply:
`cb` should probably be moved in here